plugins/omjournal/omjournal.c

```diff
@@ -181,111 +181,105 @@ ENDtryResume
 
 
 struct iovec *
-build_iovec(size_t *argc, struct json_object *json) {
-
-    struct lh_entry *entry ;
-    struct iovec *iov;
-    const char *key;
-    const char *val;
-    size_t key_len;
-    size_t val_len;
-    size_t vec_len;
-    size_t i;
-
-    *argc = json_object_object_length(json);
-    iov = malloc( sizeof(struct iovec) * *argc );
-    entry = json_object_get_object(json)->head;
-
-    if(NULL == iov)
-        goto fail;
-
-    /* 
-     * we have to avoid using json_object_object_foreachC because clang static analyzer doesn't believe
-     * that I've correct initialised all the elements of iov.
-     * I'm assuming that json_object_object_length isn't lying to me, and that the json object isn't
-     * changing under my feet so that we can do an explicit `for` iteration instead of just walking the
-     * the linked list.
-     */
-
-    for(i = 0; i < *argc; i++)
-    {
-        key = (char *)entry->k;
-        val = json_object_get_string((struct json_object*)entry->v);
-
-        key_len = strlen(key);
-        val_len = strlen(val);
-        // vec length is len(key=val)
-        vec_len = key_len + val_len + 1;
-
-        char *buf = malloc(vec_len + 1);
-        if(NULL == buf) 
-            goto fail;
-       
-        memcpy(buf, key, key_len);
-        memcpy(buf + key_len, "=", 1);
-        memcpy(buf + key_len + 1, val, val_len+1);
-
-        iov[i].iov_base = buf;
-        iov[i].iov_len = vec_len;
-
-        entry = entry->next;
-    }
-    return iov;
+build_iovec(size_t *argc, struct json_object *json)
+{
+	struct iovec *iov;
+	const char *key;
+	const char *val;
+	size_t key_len;
+	size_t val_len;
+	size_t vec_len;
+	size_t i;
+
+	*argc = json_object_object_length(json);
```

Comment:
Would it be worth making sure we are going to be successful before we write to external arguments?  For example, if we have a number of elements, but memory allocation fails, we have written a positive number to `*argc`, but then are returning NULL.  Perhaps instead we should write `*argc` when just before we `return`?